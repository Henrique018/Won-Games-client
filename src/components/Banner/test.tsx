import { screen } from '@testing-library/react';
import { renderWithTheme } from 'utils/test/helper';

import Banner, { BannerProps } from '.';

const args = {
	img: 'https://source.unsplash.com/user/willianjusten/1042x580',
	title: 'Defy death',
	subtitle: '<p>Play the new <strong>CrashLands</strong> season',
	buttonLabel: 'Buy now',
	buttonLink: '/games/defy-death',
} as BannerProps;
// todo banner tests..
describe('<Banner />', () => {
	it('should render the banner correctly', () => {
		const { container } = renderWithTheme(<Banner {...args} />);

		expect(screen.getByRole('img')).toHaveAttribute('src');

		expect(
			screen.getByRole('heading', { name: /Defy death/i })
		).toBeInTheDocument();

		expect(screen.getByRole('link')).toBeInTheDocument();

		expect(container.firstChild).toMatchSnapshot();
	});
});
