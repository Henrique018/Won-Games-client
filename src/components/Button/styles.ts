import styled, { css, DefaultTheme } from 'styled-components';
import { ButtonProps } from '.';

type WrapperProps = { hasIcon: boolean } & Pick<
	ButtonProps,
	'size' | 'fullWidth'
>;

const wrapperModifier = {
	large: (theme: DefaultTheme) => css`
		height: 5rem;
		width: 15rem;
		font-size: ${theme.font.sizes.medium};
		padding: ${theme.spacings.xxsmall} ${theme.spacings.large};
	`,

	medium: (theme: DefaultTheme) => css`
		height: 4rem;
		width: 12.2rem;
		font-size: ${theme.font.sizes.small};
		padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
	`,

	small: (theme: DefaultTheme) => css`
		height: 3rem;
		width: 9.5rem;
		font-size: ${theme.font.sizes.xsmall};
	`,

	fullWidth: () => css`
		width: 100%;
	`,
	withIcon: (theme: DefaultTheme) => css`
		display: flex;
		align-items: center;
		justify-content: center;
		svg {
			width: 1.5rem;

			& + span {
				margin-left: ${theme.spacings.xxsmall};
			}
		}
	`,
};

export const Wrapper = styled.button<WrapperProps>`
	${({ theme, size, fullWidth, hasIcon }) => css`
		color: ${theme.colors.white};
		cursor: pointer;
		border: 0;
		border-radius: ${theme.border.radius};
		padding: ${theme.spacings.xxsmall};
		background: linear-gradient(
			180deg,
			#ff5f5f -14.51%,
			#f062c0 102.86%,
			#f23131 102.86%
		);

		&:hover {
			background: linear-gradient(180deg, #e35565 0%, #d958a6 100%),
				linear-gradient(
					180deg,
					#ff5f5f -14.51%,
					#f062c0 102.86%,
					#f23131 102.86%
				);
		}

		${!!size && wrapperModifier[size](theme)}
		${!!fullWidth && wrapperModifier.fullWidth()}
		${hasIcon && wrapperModifier.withIcon(theme)}
	`}
`;
