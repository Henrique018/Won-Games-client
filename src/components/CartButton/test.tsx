import { render, screen } from 'utils/test-util';
import userEvent from '@testing-library/user-event';

import { cartContextDefaultValues } from 'hooks/use-cart';

import CartButton from '.';

describe('<CartButton />', () => {
	it('should render and execute addToCart method if clicked', () => {
		const cartProviderProps = {
			...cartContextDefaultValues,
			isInCart: () => false,
			addToCart: jest.fn(),
		};

		render(<CartButton id="1" />, { cartProviderProps });

		const button = screen.getByLabelText(/add to cart/);

		expect(button).toBeInTheDocument();

		userEvent.click(button);

		expect(cartProviderProps.addToCart).toHaveBeenCalledWith('1');
	});

	it('should execute removeFromCart method if clicked', () => {
		const cartProviderProps = {
			...cartContextDefaultValues,
			isInCart: () => true,
			addToCart: jest.fn(),
			removeFromCart: jest.fn(),
		};

		render(<CartButton id="1" />, { cartProviderProps });

		const button = screen.getByLabelText(/remove from cart/);
		userEvent.click(button);
		expect(cartProviderProps.removeFromCart).toHaveBeenCalledWith('1');
	});
});
