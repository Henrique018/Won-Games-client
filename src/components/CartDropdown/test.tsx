import { cartContextDefaultValues } from 'hooks/use-cart';
import { render, screen } from 'utils/test-util';
import cartMock from 'components/CartList/mock';

import CartDropdown from '.';

describe('<CartDropdown />', () => {
	beforeEach(() => {
		const cartProviderProps = {
			...cartContextDefaultValues,
			quantity: cartMock.items.length,
			items: cartMock.items,
			total: '$335,00',
		};

		render(<CartDropdown />, { cartProviderProps });
	});

	it('should render CartIcon and its badge', () => {
		expect(screen.getByLabelText(/cart items/i)).toBeInTheDocument();
		expect(screen.getByLabelText(/shopping cart/i)).toBeInTheDocument();
	});

	it('should render a dropdown with cart items', () => {
		expect(screen.getByText(`$335,00`)).toBeInTheDocument();
		expect(screen.getByText(`${cartMock.items[0].title}`)).toBeInTheDocument();
	});
});
