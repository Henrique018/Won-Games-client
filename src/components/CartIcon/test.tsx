import { cartContextDefaultValues } from 'hooks/use-cart';

import { render, screen } from 'utils/test-util';

import CartIcon from '.';
describe('<CartIcon />', () => {
	it('should render without badge', () => {
		render(<CartIcon />);
		expect(screen.getByLabelText(/shopping cart/i)).toBeInTheDocument();
		expect(screen.queryByLabelText(/cart items/i)).not.toBeInTheDocument();
	});

	it('should render with badge', () => {
		render(<CartIcon />, {
			cartProviderProps: { ...cartContextDefaultValues, quantity: 12 },
		});
		expect(screen.queryByLabelText(/cart items/i)).toBeInTheDocument();
		expect(screen.queryByText(/12/i)).toBeInTheDocument();
	});
});
