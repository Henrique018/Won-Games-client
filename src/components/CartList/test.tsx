import { screen } from '@testing-library/react';
import { renderWithTheme } from 'utils/test/helper';

import CartList from '.';

import cartMock from './mock';

describe('<CartList />', () => {
	it('should render correctly', () => {
		const { container } = renderWithTheme(<CartList {...cartMock} />);

		expect(
			screen.getByRole('heading', { name: 'Borderlands 3' })
		).toBeInTheDocument();
		expect(
			screen.getByRole('heading', { name: 'Red Dead Redemption 2' })
		).toBeInTheDocument();
		expect(screen.getByText(/total/i)).toBeInTheDocument();
		expect(screen.getByText('R$ 335,00')).toHaveStyle({ color: '#F231A5' });

		expect(container.firstChild).toMatchSnapshot();
	});

	it('should render a button', () => {
		renderWithTheme(<CartList {...cartMock} hasButton />);

		expect(screen.getByRole('link', { name: /buy it now!/i }));
	});

	it('should render the empty component if there are no games', () => {
		renderWithTheme(<CartList items={[]} hasButton />);

		expect(screen.getByText(/Your cart is empty/i)).toBeInTheDocument();
		expect(screen.queryByText(/total/i)).not.toBeInTheDocument();
	});
});
