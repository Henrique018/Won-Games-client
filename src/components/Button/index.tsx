import { forwardRef, AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react';
import * as S from './styles';

type ButtonTypes =
	| AnchorHTMLAttributes<HTMLAnchorElement>
	| ButtonHTMLAttributes<HTMLButtonElement>;

export type ButtonProps = {
	size?: 'large' | 'medium' | 'small';
	fullWidth?: boolean;
	icon?: React.ReactNode;
	as?: React.ElementType;
	minimal?: boolean;
} & ButtonTypes;

const Button: React.ForwardRefRenderFunction<S.WrapperProps, ButtonProps> = (
	{
		children,
		size = 'medium',
		fullWidth = false,
		icon,
		minimal = false,
		...props
	},
	ref
) => {
	return (
		<S.Wrapper
			size={size}
			fullWidth={fullWidth}
			hasIcon={!!icon}
			minimal={minimal}
			ref={ref}
			{...props}
		>
			{!!icon && icon}
			{!!children && <span>{children}</span>}
		</S.Wrapper>
	);
};

export default forwardRef(Button);
