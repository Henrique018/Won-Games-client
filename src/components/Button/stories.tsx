import { Story, Meta } from '@storybook/react/types-6-0';
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart';

import Button from '.';

export default {
	title: 'Button',
	component: Button,
	argTypes: {
		icon: { type: '' },
	},
} as Meta;

export const Default: Story = (args) => <Button {...args} />;
export const WithIcon: Story = (args) => <Button {...args} />;

Default.args = {
	children: 'Buy now',
};

WithIcon.args = {
	children: 'Buy now',
	icon: <AddShoppingCart />,
};
