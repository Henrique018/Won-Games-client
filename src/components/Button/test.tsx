import { screen } from '@testing-library/react';
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart';

import { renderWithTheme } from 'utils/test/helper';
import Button from '.';

describe('<Button />', () => {
	it('should render a medium size button by default', () => {
		const { container } = renderWithTheme(<Button>Download Now</Button>);

		expect(screen.getByRole('button')).toHaveStyle({
			width: '12.2rem',
			height: '4rem',
			padding: '0.8rem 3.2rem',
			'font-size': '1.4rem',
		});

		expect(container.firstChild).toMatchSnapshot();
	});

	it('should render a small size button', () => {
		renderWithTheme(<Button size="small">Download Now</Button>);

		expect(screen.getByRole('button')).toHaveStyle({
			width: '9.5rem',
			height: '3rem',
			'font-size': '1.2rem',
		});
	});

	it('should render a large size button', () => {
		renderWithTheme(<Button size="large">Download Now</Button>);

		expect(screen.getByRole('button')).toHaveStyle({
			width: '15rem',
			height: '5rem',
			padding: '0.8rem 4.0rem',
			'font-size': '1.6rem',
		});
	});

	it('should render a full width button', () => {
		renderWithTheme(<Button fullWidth>Download Now</Button>);

		expect(screen.getByRole('button')).toHaveStyle({
			width: '100%',
		});
	});

	it('should render an icon version', () => {
		renderWithTheme(
			<Button icon={<AddShoppingCart data-testid="icon" />}>
				Download Now
			</Button>
		);

		expect(screen.getByText(/download now/i)).toBeInTheDocument();
		expect(screen.getByTestId('icon')).toBeInTheDocument();
	});

	it('should render the button as a link', () => {
		const { debug, container } = renderWithTheme(
			<Button as="a" href="/link">
				Buy now
			</Button>
		);

		debug(container);
		expect(screen.getByRole('link', { name: /buy now/i })).toHaveAttribute(
			'href',
			'/link'
		);
	});
});
