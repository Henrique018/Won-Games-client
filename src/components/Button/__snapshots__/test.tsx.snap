// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render a medium size button by default 1`] = `
.c0 {
  color: #FAFAFA;
  cursor: pointer;
  border: 0;
  border-radius: 0.4rem;
  padding: 0.8rem;
  background: linear-gradient( 180deg,#ff5f5f -14.51%,#f062c0 102.86%,#f23131 102.86% );
  height: 4rem;
  width: 12.2rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

<button
  class="c0"
>
  <span>
    Download Now
  </span>
</button>
`;
